crates/holochain/src/core/workflow/incoming_dht_ops_workflow.rs

```diff
@@ -2,109 +2,87 @@
+        let mut set = incoming_op_hashes.0.lock();
 
-impl Default for IncomingOpsBatch {
-    fn default() -> Self {
-        Self(RwShare::new(InOpBatch::default()))
-    }
-}
+        // Track the hashes that we're going to work on, and should be removed from the shared state
+        // when this claim is dropped.
+        let mut working_hashes = Vec::with_capacity(ops.len());
+        let mut working_ops = Vec::with_capacity(ops.len());
 
-/// if result.0.is_none() -- we queued it to send later
-/// if result.0.is_some() -- the batch should be run now
-fn batch_check_insert(
-    batch: &IncomingOpsBatch,
-    request_validation_receipt: bool,
-    ops: Vec<(DhtOpHash, DhtOp)>,
-) -> (Option<Vec<InOpBatchEntry>>, InOpBatchRcv) {
-    let (snd, rcv) = tokio::sync::oneshot::channel();
-    let entry = InOpBatchEntry {
-        snd,
-        request_validation_receipt,
-        ops,
-    };
-    batch.0.share_mut(|batch| {
-        if batch.is_running {
-            // there is already a batch running, just queue this
-            batch.pending.push(entry);
-            (None, rcv)
-        } else {
-            // no batch running, run this (and assert we never collect straglers
-            assert!(batch.pending.is_empty());
-            batch.is_running = true;
-            (Some(vec![entry]), rcv)
+        for op in ops {
+            if !set.contains(&op.hash) {
+                set.insert(op.hash.clone());
+                working_hashes.push(op.hash.clone());
+                working_ops.push(op);
+            }
         }
-    })
+
+        (
+            Self {
+                incoming_op_hashes: keep_incoming_op_hashes,
+                working_hashes,
+            },
+            working_ops,
+        )
+    }
 }
 
-/// if result.is_none() -- we are done, end the loop for now
-/// if result.is_some() -- we got more items to process
-fn batch_check_end(batch: &IncomingOpsBatch) -> Option<Vec<InOpBatchEntry>> {
-    batch.0.share_mut(|batch| {
-        assert!(batch.is_running);
-        let out: Vec<InOpBatchEntry> = batch.pending.drain(..).collect();
-        if out.is_empty() {
-            // pending was empty, we can end the loop for now
-            batch.is_running = false;
-            None
-        } else {
-            // we have some more pending, continue the running loop
-            Some(out)
+impl Drop for OpsClaim {
+    fn drop(&mut self) {
+        // Lock the shared state while we remove the ops we're finished working with
+        let incoming_op_hashes = self.incoming_op_hashes.clone();
+        let mut set = incoming_op_hashes.0.lock();
+
+        for hash in &self.working_hashes {
+            set.remove(hash);
         }
-    })
+    }
 }
 
 #[instrument(skip(txn, ops))]
 fn batch_process_entry(
     txn: &mut rusqlite::Transaction<'_>,
     request_validation_receipt: bool,
-    ops: Vec<(DhtOpHash, DhtOp)>,
+    ops: Vec<DhtOpHashed>,
```

Comment:
We were hashing before the workflow call AND right before inserting into the database. That means the workflow trusts its caller but then computes new hashes just to be safe? It seems to me that it makes more sense for hashing to take place at the top of the workflow so that it is self-contained and clearly maintains its own integrity. Then there's no need to trust the caller and no need to hash twice.

This removes the risk of claiming a different hash than we're actually going to end up inserting into the database if the workflow gets called with a wrong hash for some reason. That would create a real mess but I don't think that was actually possible in the existing code.